### ISSUE TITLE
lazyConnect should not affect api executions

### ISSUE THREAD

**Issue by jannyHou** [strongloop/loopback#2850]
### Bug or feature request
- [x] Bug
- [ ] Feature request
### Description of feature (or steps to reproduce if bug)

`lazyConnect` should only defer the connection phase, but connection will still establish when users call endpoints.
Now if we set `lazyConnect = true` in datasource config, the db instance will not be attached to datasource and all endpoints get 500 server error.
### Link to sample repo to reproduce issue (if bug)
1. create a mongodb datasource with `"lazyConnect": true`
2. boot the app, open explorer
### Expected result
- [x] The connection is deferred when booting app
- [ ] apis in explorer should still work (not working now)
### Actual result (if bug)

apis in explorer not work.

**Comment by jannyHou** [comment]
The related doc is removed to prevent users using it.
The actual fix is TBD

**Comment by manjubhat** [comment]
I am trying to use lazyConnect along with the recommendation in   ##https://github.com/strongloop/loopback-connector-mysql/issues/118#issuecomment-129928045 to run ds.autoupdate (To create database tables for the first time). But first api execution always fails with  500 Internal server error. It seems to be because lazyConnect triggers connection to DB when first api request is made. On 'connected'  event, if we performing ds.autoupdate, API execution query also happens on same event, but it may happen before table creation is complete. Is this a bug ?

@raymondfeng ,
Do you have any recommendation on how to run ds.autoupdate when lazyConnect is used without affecting first API execution

**Comment by bajtos** [comment]
> Do you have any recommendation on how to run ds.autoupdate when lazyConnect is used without affecting first API execution

@manjubhat could you please open a new issue to discuss this particular problem? Also it would be great to have either a small app we can use to reproduce the issue locally (see https://loopback.io/doc/en/contrib/Reporting-issues.html#loopback-3x-bugs) or even better a test case for loopback-datasource-juggler or loopback-connector-mysql.

@jannyHou could you please take a look at this problem too?

**Comment by bajtos** [comment]
> could you please open a new issue to discuss this particular problem?

I take that back. Let's keep the discussion in this issue for now, I cannot tell whether it's the same problem as described in the original description or something different.

**Comment by jannyHou** [comment]
@manjubhat Can you try to call method `connect` or `ping` as the first call?

**Comment by manjubhat** [comment]
I tried connect as well as ping as the first call. When the DB is not available, connect/ping hangs forever. It seems DB should be ready to accept connections when we call connect. That may not be always the case for us.

One possible solution is to implement DB Availability check using socket.connect and retries. Call connect only after DB is ready to accept connections. This way we have data source connected and tables created before the first api execution. @jannyHou , Let me know if there any other ways to connect to database which include retries.
